### PR TITLE
 Implement blockchain.Finalize 

### PR DESCRIPTION
**Pull request by omerfirmak**

### CODE REVIEW

File: sync/sync_test.go [hunk 1]
```diff
@@ -115,7 +115,7 @@ func TestSyncBlocks(t *testing.T) {
 			case 2:
 				state.BlockHash = new(felt.Felt) // fail sanity checks
 			case 3:
-				state.OldRoot = new(felt.Felt).SetUint64(1) // fail store
+				state.NewRoot = new(felt.Felt).SetUint64(1) // fail store
```
Comment (joshklop) [inline sync/sync_test.go:118]: What is the reason for this change?

File: core/block.go [hunk 2]
```diff
@@ -150,6 +150,14 @@ func VerifyBlockHash(b *Block, network utils.Network) (*BlockCommitments, error)
 	return nil, errors.New("can not verify hash in block header")
 }
 
+// BlockHash computes the block hash
+func BlockHash(b *Block, network utils.Network) (*felt.Felt, *BlockCommitments, error) {
+	if b.Number < NetworkBlockHashMetaInfo(network).First07Block {
+		return pre07Hash(b, network.ChainID())
+	}
+	return post07Hash(b, nil)
```
Comment (joshklop) [inline core/block.go:158]: We can export `blockHash` or do:
```go
return blockHash(b, network, nil)
```

File: blockchain/blockchain.go [hunk 3]
```diff
@@ -336,47 +336,50 @@ func (b *Blockchain) Store(block *core.Block, blockCommitments *core.BlockCommit
...
 		}
 
-		state := core.NewState(txn)
-		if err := state.Update(block.Number, stateUpdate.StateDiff, newClasses); err != nil {
+		if err := core.NewState(txn).Update(block.Number, stateUpdate.StateDiff, newClasses); err != nil {
 			return err
 		}
 
-		if err := b.verifyStateUpdateRoot(state, stateUpdate.NewRoot); err != nil {
+		if err := b.verifyStateUpdateRoot(core.NewState(txn), stateUpdate.NewRoot); err != nil {
```
Comment (joshklop) [inline blockchain/blockchain.go:347]: Is there a reason to use two different state instances?

File: blockchain/blockchain_test.go [hunk 4]
```diff
@@ -797,3 +797,47 @@ func TestMakeStateDiffForEmptyBlock(t *testing.T) {
...
+	require.NoError(t, err)
+	require.Equal(t, uint64(0), h)
+
+	pending, err := chain.Pending()
+	require.NoError(t, err)
+	require.NoError(t, chain.Finalize(&pending))
+	h, err = chain.Height()
+	require.NoError(t, err)
+	require.Equal(t, uint64(1), h)
+}
```
Comment (joshklop) [inline blockchain/blockchain_test.go:843]: Can we test that the header, txs, and receipts have been stored correctly as well?

File: core/block.go (hunk 2 above)
Comment (omerfirmak) [reply core/block.go:158]: Nice, thanks!

File: blockchain/blockchain.go (hunk 3 above)
Comment (omerfirmak) [reply blockchain/blockchain.go:347]: No real reason. Since `core.State` is just a wrapper around a `db.Transaction` it is safe to assume all changes are persisted in the `txn`. But will change it to use a single state instance to make easier to follow.

File: blockchain/blockchain_test.go (hunk 4 above)
Comment (omerfirmak) [reply blockchain/blockchain_test.go:843]: Done, that actually uncovered a bug when unmarshaling empty blocks from storage:D

File: sync/sync_test.go (hunk 1 above)
Comment (omerfirmak) [reply sync/sync_test.go:118]: I took the liberty to remove OldRoot check, since it is kinda redundant. `verifyBlock` already checks the parent block hash against the local `HEAD`, which already proves that our state root is `OldRoot`

**Review by joshklop — APPROVED** [review]
